### PR TITLE
Galactic update

### PR DESCRIPTION
**Pull request by jacobperron**
Changes so that we can build and test against Galactic (the latest ROS 2 distro).

This includes ports of the following changes from the OSRF fork:

- https://github.com/osrf/ros2_java/commit/e1eeef035dd22c5099c4d6a184e8fa50487fa301
- https://github.com/osrf/ros2_java/pull/47
- https://github.com/osrf/ros2_java/pull/63

### CODE REVIEW

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -55,7 +55,7 @@ jobs:
         cd ros2_java_ws
         curl -sL file://${{ github.workspace }}/ros2_java_android.repos | vcs import src
         # Use checked out version of ros2_java
-        rm -rf src/ros2_java/ros2_java
+        rm -rf src/ros2-java/ros2_java
         ln --symbolic ${{ github.workspace }} src/ros2_java
```
Comment (ivanpauno) [inline .github/workflows/build_and_test.yml:59]: why the change, it doesn't same to match the symbolik link done next line

File: rcljava/include/org_ros2_rcljava_client_ClientImpl.h [hunk 2]
```diff
@@ -25,9 +25,9 @@ extern "C" {
  * Method:    nativeSendClientRequest
  * Signature: (JJJJJLorg/ros2/rcljava/interfaces/MessageDefinition;)V
  */
-JNIEXPORT void
+JNIEXPORT jlong
 JNICALL Java_org_ros2_rcljava_client_ClientImpl_nativeSendClientRequest(
-  JNIEnv *, jclass, jlong, jlong, jlong, jlong, jlong, jobject);
+  JNIEnv *, jclass, jlong, jlong, jlong, jlong, jobject);
```
Comment (ivanpauno) [inline rcljava/include/org_ros2_rcljava_client_ClientImpl.h:30]: is this backporting a change from osrf/ros2_java?

if not, it will introduce some unfortunate conflicts

Comment (ivanpauno) [reply rcljava/include/org_ros2_rcljava_client_ClientImpl.h:30]: based on the commits, I see you're actually backporting some changes.
can you update the description and link the PRs (or commits) that are being ported?

We can better keep track of things that way

Comment (jacobperron) [reply rcljava/include/org_ros2_rcljava_client_ClientImpl.h:30]: Yes, I ported a few select changes towards getting CI green for Galactic. Sorry for picking things out of order, but I figured it would be better to get CI green for Galactic before porting other changes.

I've updated the PR description with references to the original changes.

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (jacobperron) [reply .github/workflows/build_and_test.yml:59]: I changed the repos file so that the path to the cloned repository matches the GitHub org where it came from (ie. ros2-java vs. ros2_java). Though it probably doesn't make a difference, I'll update the symlink for consistency.

File: rcljava/include/org_ros2_rcljava_client_ClientImpl.h (hunk 2 above)
Comment (jacobperron) [reply rcljava/include/org_ros2_rcljava_client_ClientImpl.h:30]: Let me know if you think there's a better strategy.

File: rosidl_generator_java/package.xml [hunk 3]
```diff
@@ -22,26 +22,27 @@
...
   <test_depend>ament_cmake_gtest</test_depend>
   <test_depend>ament_lint_auto</test_depend>
   <test_depend>ament_lint_common</test_depend>
 
-  <test_depend>rmw_implementation</test_depend>
   <test_depend>rmw_implementation_cmake</test_depend>
-  <test_depend>rosidl_generator_c</test_depend>
+  <!-- It's not clear why we need to depend on rosidl_generator_cpp, but without it
+       the tests will not compile.
+       See https://github.com/ros2/rosidl_python/pull/100#issuecomment-608558735 -->
```
Comment (ivanpauno) [inline rosidl_generator_java/package.xml:36]: this is interesting ...

File: .github/workflows/build_and_test.yml [hunk 4]
```diff
@@ -4,24 +4,24 @@ on: [push, pull_request]
...
     - name: Install Java
       run: |
         sudo apt-get update -qq
         sudo apt-get install -y default-jdk gradle
     - uses: actions/checkout@v2
-    - uses: ros-tooling/setup-ros@0.2.1
+    - uses: ros-tooling/setup-ros@v0.2
       with:
-        required-ros-distributions: dashing
+        required-ros-distributions: galactic
```
Comment (ivanpauno) [inline .github/workflows/build_and_test.yml:16]: this will introduce conflicts with changes in https://github.com/osrf/ros2_java/

File: ros2_java_desktop.repos [hunk 5]
```diff
@@ -1,33 +1,33 @@
 repositories:
-  ament/ament_java:
-    type: git
-    url: https://github.com/ros2-java/ament_java.git
-    version: main
   ros2/common_interfaces:
     type: git
     url: https://github.com/ros2/common_interfaces.git
-    version: dashing
+    version: galactic
```
Comment (ivanpauno) [inline ros2_java_desktop.repos:5]: this will introduce conflicts as well

File: rosidl_generator_java/cmake/rosidl_generator_java_generate_interfaces.cmake [hunk 6]
```diff
@@ -71,8 +71,12 @@ foreach(_abs_idl_file ${rosidl_generate_interfaces_ABS_IDL_FILES})
       "${_output_path}/${_parent_folder}/${_idl_name}_Response.java"
     )
     foreach(_typesupport_impl ${_typesupport_impls})
-      list(APPEND _generated_extension_${_typesupport_impl}_files "${_output_path}/${_parent_folder}/${_idl_name}_Request.ep.${_typesupport_impl}.cpp")
-      list(APPEND _generated_extension_${_typesupport_impl}_files "${_output_path}/${_parent_folder}/${_idl_name}_Response.ep.${_typesupport_impl}.cpp")
+      list(APPEND _generated_extension_${_typesupport_impl}_files
```
Comment (ivanpauno) [inline rosidl_generator_java/cmake/rosidl_generator_java_generate_interfaces.cmake:74]: this changes will introduce confilcts with https://github.com/osrf/ros2_java/pull/73 if not imported from that commit

File: rcljava/include/org_ros2_rcljava_client_ClientImpl.h (hunk 2 above)
Comment (ivanpauno) [reply rcljava/include/org_ros2_rcljava_client_ClientImpl.h:30]: The strategy sounds fine

File: .github/workflows/build_and_test.yml (hunk 4 above)
Comment (ivanpauno) [reply .github/workflows/build_and_test.yml:16]: this conflicts don't matter much, I'm ok with them

File: ros2_java_desktop.repos (hunk 5 above)
Comment (ivanpauno) [reply ros2_java_desktop.repos:5]: I'm also okay with this one

File: rosidl_generator_java/cmake/rosidl_generator_java_generate_interfaces.cmake (hunk 6 above)
Comment (ivanpauno) [reply rosidl_generator_java/cmake/rosidl_generator_java_generate_interfaces.cmake:74]: I prefer to import https://github.com/osrf/ros2_java/pull/73 if possible, which doesn't seem to be the case.

Comment (jacobperron) [reply rosidl_generator_java/cmake/rosidl_generator_java_generate_interfaces.cmake:74]: I tried cherry-picking  osrf#73 but it doesn't apply cleanly. It seemed easier to directly address the cmake warnings here..

**Comment by jacobperron** [comment]
I'm looking into the Android CI failure.

**Review by ivanpauno — APPROVED** [review]

**Comment by mclayton7** [comment]
@jacobperron I started looking into the Android CI issue and it seems to be related to https://github.com/foonathan/memory/pull/60. I'm not a ROS2/Ament expert, but it looks like a cross compilation issue that may be resolved by pointing cmake to qemu via CMAKE_CROSSCOMPILING_EMULATOR

**Comment by ivanpauno** [comment]
> @jacobperron I started looking into the Android CI issue and it seems to be related to foonathan/memory#60. I'm not a ROS2/Ament expert, but it looks like a cross compilation issue that may be resolved by pointing cmake to qemu via CMAKE_CROSSCOMPILING_EMULATOR

Thanks, let's see if https://github.com/ros2-java/ros2_java/pull/187/commits/9a781803f3af2a684241f2794d87f72dba2c3fd3 fixes the issue.

**Comment by ivanpauno** [comment]
> Thanks, let's see if 9a78180 fixes the issue.

That seem to have solved the `foonathan_memory_vendor` compile error, now there's a different issue with `rosidl_runtime_cpp`.

**Comment by ivanpauno** [comment]
> Cloning into 'mimick-f171450b5ebaa3d2538c762a059dfc6ab7a01039'...
  HEAD is now at f171450 Add armv7l as a 32-bit ARM architecture. (#16)
  CMake Error at CMakeLists.txt:88 (message):
  Architecture 'armv7-a' is not supported.

That one is unfortunate ....
IMO, we should make mimick tests optional, and disable them when mimick is not present.

**Comment by mclayton7** [comment]
> > Cloning into 'mimick-f171450b5ebaa3d2538c762a059dfc6ab7a01039'...
> > HEAD is now at f171450 Add armv7l as a 32-bit ARM architecture. (#16)
> > CMake Error at CMakeLists.txt:88 (message):
> > Architecture 'armv7-a' is not supported.
> 
> That one is unfortunate ....
> IMO, we should make mimick tests optional, and disable them when mimick is not present.

It looks like https://github.com/ros2/Mimick supports ARM/ARM64, so maybe it's just a matter of tweaking the cmake scripts in https://github.com/ros2/mimick_vendor? I'll try to look at it this weekend.

**Comment by ivanpauno** [comment]
> It looks like https://github.com/ros2/Mimick supports ARM/ARM64, so maybe it's just a matter of tweaking the cmake scripts in https://github.com/ros2/mimick_vendor? I'll try to look at it this weekend.

Android uses a different ABI, so I'm not sure if it will work, probably it won't.
Mimick is a C mocking library that works by poisoning the GOT of a dynamic library at runtime, so it's pretty platform dependent.
IMO, it should be possible to disable all tests using mimick in the ros2 core by passing a cmake flag, I will see what I can do.

We can also just not build tests at all.... but that doesn't sound like a good idea.

**Comment by ivanpauno** [comment]
I have opened a PR in rcl to skip mimick tests https://github.com/ros2/rcl/pull/923.
We will need to do similar work in `rcutils`, `rclcpp` and in the other packages in the `rcl` repo, and that will likely fix the android builds.

**Comment by ivanpauno** [comment]
The last five commits skipped building tests on the packages using mimick, so we are not blocked on PRs in the ros2 core to merge this.

Now this is failing in the rcutils shared library helpers ....
This is getting fun :smile:

**Comment by ivanpauno** [comment]
@jacobperron CI is passing now. 
The only job that failed is duplicated (push and pr triggers), and had a random issue.

**Comment by jacobperron** [comment]
@ivanpauno Thanks for looking into it! I think we can merge this now. Maybe we can remove the duplicate jobs in a follow-up.

**Comment by jacobperron** [comment]
I've tried to cleanup the commits a little bit; I'll avoid squashing to preserve history.
